added ISO15693 (NfcV) reading, saving, emulating and revealing from privacy mode (unlock)

g3gg0 opened the pull request:
# What's new

This PR is related to #1991 (without all its history), built on top of #2070 (candidate for merge) and from a new branch.

- Added NfcV through various layers
- Added NfcV memory and system info reading
- Added loading/saving/emulating
- Added unlocking NfcV type NXP SLIX-L (reveal from privacy mode)
- Usable for another module already in development by @nvx (picopass emulation)


This is a PR, being aware that:
- rfalNfcv* calls are being made from lib/ level
- Nfc code is undergoing a refactoring

Q: Might someone of the maintainers help touching some of the quesionable code sections, as i am still not very confident with
the inner workings etc?

I consider the code being good enough for a PR process, few reworking will have to be done. I you disagree, please tell.

regards,
g3gg0

g3gg0 committed:
added support for ISO15693 (NfcV) emulation, added support for reading SLIX tags

g3gg0 committed:
SLIX: fixed crash situation when an invalid password was requested

g3gg0 committed:
ISO15693: show emulate menu when opening file

g3gg0 committed:
rename NfcV emulate scene to match other NfcV names

g3gg0 committed:
optimize allocation size for signals

g3gg0 committed:
ISO15693: further optimizations of allocation and free code

g3gg0 committed:
ISO15693: further cleanup

g3gg0 committed:
ISO15693: reduce latency on state machine reset

g3gg0 committed:
further code cleanups

nemanjan00 commented:
READ_MULTI_BLOCKS seems to be broken:

```
Flipper:

        384 |      59264 | Rdr |62  23  ff  ff  ff  ff  00  00  07  e0  00  05  aa  eb                   |  ok | READ_MULTI_BLOCK(0-5)
      63296 |     177984 | Tag |00  12  34  56  78  90  12  34  56  78  90  12  34  56  78  90  12  34   |     | 
            |            |     |56  78  90  12  34  56  78  04  84                                       |  ok | 

Real card:
        384 |      59264 | Rdr |62  23  ff  ff  ff  ff  00  00  07  e0  00  05  aa  eb                   |  ok | READ_MULTI_BLOCK(0-5)
      63552 |     202816 | Tag |00  00  12  34  56  78  00  90  12  34  56  00  78  90  12  34  00  56   |     | 
            |            |     |78  90  12  00  34  56  78  90  00  12  34  56  78  55  31               |  ok | 
```

g3gg0 commented:
thanks, didn't notice and will have a look at it.

g3gg0 commented:
this is in unadressed mode:
![grafik](https://user-images.githubusercontent.com/479231/213586855-e231f9e4-cb17-4075-9a08-1665163d6909.png)

here in adressed mode:
![grafik](https://user-images.githubusercontent.com/479231/213586962-dea4044b-b99b-4b40-a30a-5b2771916d08.png)

here with 1oo256 mode:
![grafik](https://user-images.githubusercontent.com/479231/213587124-22c35e1c-7f67-4fd8-9dfc-7c921104721e.png)

why is the custom bit set in your request?

nemanjan00 commented:
![20230120_00h53m48s_grim](https://user-images.githubusercontent.com/448151/213587860-ecb353dc-5957-4fc8-b5b6-2b415a3c3cdb.png)

Looks like you are missing block security status

g3gg0 commented:
from which datasheet are you citing?

nemanjan00 commented:
BS_ISO_IEC_15693-3-2009__A2-2015.pdf

g3gg0 commented:
ah okay i have a quite old standard revision.
seems the block security status is requested when the custom flag is set (?)

need to find a latest datasheet. if someone knows where to find it -> git - at - g3gg0.de

nemanjan00 commented:
Guessing you will probably write it better, but, anyway...

```patch

diff --git a/lib/nfc/protocols/nfcv.c b/lib/nfc/protocols/nfcv.c
index cc7f4ad5..2ffdd0ea 100644
--- a/lib/nfc/protocols/nfcv.c
+++ b/lib/nfc/protocols/nfcv.c
@@ -537,6 +537,29 @@ void nfcv_emu_handle_packet(
             nfcv_emu_send(
                 tx_rx, nfcv_data, ctx->response_buffer, 1, ctx->response_flags, ctx->send_time);
         } else {
+            if(ctx->command == ISO15693_READ_MULTI_BLOCK) {
+                ctx->response_buffer[0] = ISO15693_NOERROR;
+
+                for(int current_block = 0; current_block < blocks; current_block++) {
+                    memcpy(
+                        &ctx->response_buffer[1 + (1 + nfcv_data->block_size) * current_block + 1],
+                        &nfcv_data->data[nfcv_data->block_size * (block + current_block)],
+                        nfcv_data->block_size);
+
+                        ctx->response_buffer[1 + (1 + nfcv_data->block_size) * current_block] = 0;
+                }
+
+                nfcv_emu_send(
+                    tx_rx,
+                    nfcv_data,
+                    ctx->response_buffer,
+                    1 + (nfcv_data->block_size + 1) * blocks,
+                    ctx->response_flags,
+                    ctx->send_time);
+
+                break;
+            }
+
             ctx->response_buffer[0] = ISO15693_NOERROR;
             memcpy(
                 &ctx->response_buffer[1],
```

g3gg0 commented:
thanks, when trying a real SLIX-L it looks like that:
![grafik](https://user-images.githubusercontent.com/479231/213589171-e754625a-d0bd-4ba5-8cad-a46922c1dc23.png)

so if the custom bit is set, it seems the card will send the security status.
can you check the standard if this bit 6 (0x40) in the flags is set, the security status is prepended?

nemanjan00 commented:
Yes, according to spec, Block security status is optional and depends on `Option_flag`

![20230120_01h11m15s_grim](https://user-images.githubusercontent.com/448151/213589832-222a5b79-07da-493e-bb8a-e3eda333b3f3.png)

Also I have sent you email with spec pdf

nemanjan00 commented:
Also, I assume TwoSubcarrier is not supported?

g3gg0 commented:
that is correct. this is not supported (yet).
see nfcv_emu_alloc() line 284
![grafik](https://user-images.githubusercontent.com/479231/213590568-138b1225-ff88-4ac0-ab65-5f3f0860544c.png)
here i allocate the waveforms for high and low data rate.
subcarrier modulation would have to get added here.

nemanjan00 commented:
```
Filetype: Flipper NFC device
Version: 3
# Nfc device type can be UID, Mifare Ultralight, Mifare Classic or ISO15693
Device type: ISO15693
# UID is common for all formats
UID: E0 07 00 00 FF FF FF FF
# Data Storage Format Identifier
DSFID: 00
# Application Family Identifier
AFI: 00
IC Reference: 01
# Number of memory blocks, usually 0 to 256
Block Count: 28
# Size of a single memory block, usually 4
Block Size: 04
Data Content: 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 34 56 78 90 12 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
# Subtype of this card (0 = ISO15693, 1 = SLIX, 2 = SLIX-S, 3 = SLIX-L, 4 = SLIX2)
Subtype: 00
# End of ISO15693 parameters
```

Just in case you need sample I used that made PM3 use that flag

g3gg0 commented:
first/third real SLIX-L, second/fourth flipper
![grafik](https://user-images.githubusercontent.com/479231/213592339-cfe62bfe-1f25-4b54-bbc9-6b57ffc133ac.png)

nemanjan00 commented:
I use:
```
hf 15 rdmulti -* -b 0 --cnt 6
```

g3gg0 committed:
respond with block security status when option flag is set

g3gg0 commented:
fixed that part, thanks a lot!

nemanjan00 commented:
Thank you!

nemanjan00 commented:
One of my readers causes Flipper to crash will NULL pointer when I open Log. I will be sniffing traffic tomorrow and sending more info

g3gg0 commented:
thanks. in the evening i will check where NULL dereference might have happened.

g3gg0 committed:
increased maximum memory size to match standard
added security status handling/load/save
added SELECT/QUIET handling
more fine grained allocation routines and checks
fix memset sizes

g3gg0 commented:
@nemanjan00 
I pushed a few changes. maybe you can check if you still see crashes?
thanks!

g3gg0 commented:
> One of my readers causes Flipper to crash will NULL pointer when I open Log. I will be sniffing traffic tomorrow and sending more info

this issue is also fixed. that was the case if the only received commands were addresses and the log was opened.

g3gg0 committed:
added "Listen NfcV Reader" to sniff traffic from reader to card

g3gg0 committed:
added correct description to delete menu

g3gg0 committed:
also added DSFID/AFI handling and locking

g3gg0 committed:
increase sniff log size

g3gg0 commented:
ok, added a few more optional commands, just to make it more complete.
not sure which systems use them, but now DSFID and AFI should be handled as the standard suggests.

g3gg0 committed:
scale NfcV frequency a bit, add echo mode, fix signal level at the end

gornekich commented:
Hello @g3gg0 !
Thanks for your work on NFC-V support! Since this PR is based on #2070 , I convert this PR to draft until your previous PR is merged. However ,I am starting to test and review this PR now

nvx committed:
fix debug print warnings

nvx commented:
Rebased onto latest version of #2070 (which in turn was rebased onto the latest dev branch). Once #2070 is merged this should be good to go as well.

g3gg0 commented:
@nvx  you tell me when to issue this PR for review again?

nvx commented:
> @nvx you tell me when to issue this PR for review again?

Will do. I figure worth leaving it as draft until #2070 is merged, then a quick rebase if necessary and can submit for review then.

g3gg0 commented:
for those who have issues with the ISO15693 emulation, there is also some issue when reading with mobile phones using e.g. "NFC Tools".
however from the looks @nvx and me are quite certain this feature sends the correct data, but the phone cannot detect the bits.
could as well be the antenna or the NFC chips modulation rate causing issues.

nvx committed:
Minor fixes

nvx committed:
update api_symbols.csv

nvx commented:
Branch updated to latest dev and api symbols updated (both f7 and f18 targets). Functionality tested fine. Ready for merge or feedback :)

nvx commented:
Branch updated to latest dev and api symbols updated (both f7 and f18 targets). Functionality tested fine. Ready for merge or feedback :)

g3gg0 committed:
use symbolic modulated/unmodulated GPIO levels

g3gg0 committed:
honor AFI field, decrease verbosity and removed debug code

g3gg0 commented:
here as reference a comparison of flipper' response (top) and an original SLIX-L (bottom) to an IDENTIFY request.

you see the reader's request last bits on the left side and the load modulation on the right side
![grafik](https://github.com/flipperdevices/flipperzero-firmware/assets/479231/1dcc1620-6c9b-4f47-99a6-bf4cbd56236d)

zoomed in a bit, shows the modulation which initially looked very promising, but zooming into it, flipper's modulation is a bit more noisy. 
might come from the general RF noise emitted by flipper, not sure. thats nothing that can (reasonably) be influenced by these patches.
![grafik](https://github.com/flipperdevices/flipperzero-firmware/assets/479231/1d719178-685a-4056-807e-89e1944d6e27)

edit:
to prevent misunderstandings: i don't think that causes any issues.
it's just the signal is not 100% as clean as those from tags.

g3gg0 commented:
when checking proxmark's emulation, they don't seem to have the modulation synchronized to the main field.
the last partial dip in field strength makes me believe so.
which means the modulation we generate looks comparable.

![grafik](https://github.com/flipperdevices/flipperzero-firmware/assets/479231/6e8c2764-7b56-40cf-80a9-652fa0e6276a)

g3gg0 committed:
refactor defines for less namespace pollution by using NFCV_ prefixes

g3gg0 committed:
correct an oversight that original cards return an generic error when addressing outside block range

g3gg0 committed:
use inverse modulation, increasing readable range significantly

g3gg0 committed:
rework and better document nfc chip initialization

g3gg0 commented:
made some updates which would probably would have come up during your review anyway @gornekich 
we are now using NFCV_ prefixes and also not using RFAL_ defines for non-rfal calls.
also changed all(?) constants to definitions

using "inverse modulation" compared to a real tag like e.g. proxmark does increase detection range.

nvx commented:
Is there anything blocking this from being merged that g3gg0 or myself need to address?

skotopes commented:
@nvx Planning to review it today

gornekich commented on applications/main/nfc/scenes/nfc_scene_nfc_data_info.c:
```diff
@@ -7,6 +7,17 @@ void nfc_scene_nfc_data_info_widget_callback(GuiButtonType result, InputType typ
+    }
+
+    return value;
+}
```
I suggest using nfc_util_bytes2num() instead of this function

gornekich commented on lib/nfc/protocols/nfcv.h:
```diff
@@ -0,0 +1,291 @@
+/* true: modulating releases load, false: modulating adds load resistor to field coil */
+#define NFCV_LOAD_MODULATION_POLARITY (false)
+
+#define NFCV_FC (13560000.0f / 0.9998f) /* MHz */
```
Could you explain where 0.9998f comes from?

g3gg0 replied:
Should be removed, yeah.
Was for compensating the frequencies being off a bit.
Could be systematic, could be xtal tolerances.

g3gg0 commented:
@skotopes hope no comment is a good sign and you are happy with the code as it is? :D

skotopes commented:
@g3gg0 @gornekich is reviewing this PR, also please merge dev into this branch

gornekich commented on lib/nfc/nfc_worker.c:
```diff
@@ -111,13 +111,254 @@ int32_t nfc_worker_task(void* context) {
+        furi_hal_nfc_ll_set_error_handling(FuriHalNfcErrorHandlingNfc);
+        furi_hal_nfc_ll_set_guard_time(FURI_HAL_NFC_LL_GT_NFCV);
+
+        furi_hal_console_printf("Detect presence\r\n");
```
Please rework with FURI_LOG_D() everywhere

gornekich commented on applications/main/nfc/scenes/nfc_scene_nfcv_emulate.c:
```diff
@@ -0,0 +1,164 @@
+};
+
+bool nfc_scene_nfcv_emulate_worker_callback(NfcWorkerEvent event, void* context) {
+    UNUSED(event);
```
delete please =)

gornekich commented on applications/main/nfc/scenes/nfc_scene_read.c:
```diff
@@ -68,6 +68,11 @@ bool nfc_scene_read_on_event(void* context, SceneManagerEvent event) {
             consumed = true;
+        } else if(event.event == NfcWorkerEventReadNfcV) {
+            notification_message(nfc->notifications, &sequence_success);
+            scene_manager_next_scene(nfc->scene_manager, NfcSceneNfcDataInfo);
```
I would add NfcSceneNfcvReadSuccess. There I would display NFC-V type, UID Block number and Block size. Everything should fit on the screen without scrolling. Also add "More" button there, and in NfcSceneNfcvMenu add "Info", which will switch to NfcSceneNfcDataInfo scene

gornekich commented on applications/main/nfc/scenes/nfc_scene_nfc_data_info.c:
```diff
@@ -15,7 +26,7 @@ void nfc_scene_nfc_data_info_on_enter(void* context) {
-       (protocol == NfcDeviceProtocolMifareClassic)) {
+       (protocol == NfcDeviceProtocolMifareClassic) || (protocol == NfcDeviceProtocolNfcV)) {
         widget_add_button_element(
             widget, GuiButtonTypeRight, "More", nfc_scene_nfc_data_info_widget_callback, nfc);
```
If you add NfcSceneNfcvReadSuccess scene which I mentioned above, there shouldn't be "More" button on Info scene. If I load Nfcv card from Saved menu, there is already Emulate option in SavedMenu scene.

gornekich reviewed (changes requested):
Great work, thanks!
Could you please make some fixes and merge dev

gornekich reviewed (changes requested):
Great work, thanks!
Could you please make some fixes and merge dev

gornekich commented:
Also in #2529 we introduce new NFC architecture. MfUltralight is almost reworked with new API, now we are working on MfClassic and MfDesfire. I think we will finish low level API in a couple of weeks and continue to add support of other protocols.

However API is not finalized yet, it would be nice to receive some comments from you. And if you are interested in reworking NFC-V support with new API in future, please let me know =)

nvx commented:
Will update with the requested changes tonight. Thanks for the review!

nvx committed:
Merge remote-tracking branch 'origin/dev' into ISO15693

nvx committed:
nfcv code review fixes

nvx commented:
Merged dev and made all the suggested fixes except the GUI changes. I'm not super familiar with the Flipper GUI stuff so that'll be a little longer for me to get that part done unless someone felt like beating me to it.

xMasterX commented:
> Merged dev and made all the suggested fixes except the GUI changes. I'm not super familiar with the Flipper GUI stuff so that'll be a little longer for me to get that part done unless someone felt like beating me to it.

Hi @nvx 

I made all requested GUI changes, and fixed crash that occurs when you exit from emulation and start it again
I don't see how I can make PR into PR, so I'm going to share two patches, please take them

[b0c322.patch](https://github.com/flipperdevices/flipperzero-firmware/files/11636215/b0c322.patch)
[2dc005.patch](https://github.com/flipperdevices/flipperzero-firmware/files/11636216/2dc005.patch)

nvx commented:
@xMasterX awesome! I'll get those patches applied shortly!

nvx committed:
Disable accidentally left on signal debug gpio output

xMasterX committed:
Improve NFCV Read/Info GUIs.

Authored by @xMasterX, committed by @nvx

xMasterX committed:
Fix crash that occurs when you exit from NFCV emulation and start it again.

Authored by @xMasterX, committed by @nvx

nvx committed:
Merge remote-tracking branch 'origin/dev' into ISO15693

nvx commented:
Patches merged and tested. Also merged in latest dev again as well. Cheers for the GUI patches and nice catch with that crash fix @xMasterX too!

I believe this closes out all the outstanding items from the review. Let me know if anything else is spotted.

nvx committed:
Remove delay from emulation loop. This improves compatibility when the reader is Android.

g3gg0 commented:
thanks @nvx @xMasterX for your great support here! :)

also @nvx for hunting down that "hmmm android doesn't read flipper that reliable anymore?" issue, which was a pain!

g3gg0 committed:
Merge branch 'dev' into ISO15693

gornekich committed:
Merge branch 'dev' into ISO15693

skotopes committed:
Merge branch 'dev' into ISO15693

skotopes committed:
Merge branch 'dev' into ISO15693

skotopes committed:
Lib: digital signal debug output pin info

skotopes commented:
@g3gg0 that was some serious journey, my congratulations.

team-orangeBlue commented:
and it's merged. cool!

g3gg0 commented:
wow, great!!
thanks a lot to you and all contributors!

eychei commented:
Hi everyone,

@g3gg0 
I am trying to emulate a Slix2 tag with a different originality signature.

This is from the datasheet:

• Originality signature:
32 byte ECC based originality signature

It is set by NXP in production and can be changed for customers. 

The Dymo label writers do have a different signature thus I can not emulate the tag.
This is a link to the forum post: https://www.eevblog.com/forum/reviews/dymo-550-thermal-printer-drm-hacking/

This is a dump of a tag with the originality signature (the signature is different for every UID it seems):

inventory: 0x01,0x5C,0x63,0x7B,0x3F,0x08,0x01,0x04,0xE0
sysInfo: 0x0F,0x5C,0x63,0x7B,0x3F,0x08,0x01,0x04,0xE0,0x01,0x3D,0x4F,0x03,0x01
signature: 0x6A,0x67,0x2D,0xAE,0x8D,0xFB,0xC7,0xC2,0x12,0x65,0x49,0x41,0x9B,0x3C,0x5F,0xAF,0xD7,0xC6,0xC0,0x97,0x33,0x43,0x99,0xE5,0xD0,0x5A,0xD9,0x71,0xFE,0x7B,0x74,0xA9
blocks:
0x03,0x0A,0x82,0xED,0x86,0x39,0x61,0xD2,0x03,0x14,0x1E,0x32,0xB6,0xCA,0x00,0x3C,0x62,0x64,0xCE,0x0F,0x33,0x30,0x32,0x35,0x36,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xFF,0x24,0x01,0x01,0x00,0x00,0x00,0x22,0x04,0x1E,0x00,0x23,0x00,0x00,0x00,0x00,0x00,0x10,0x00,0xF8,0x03,0x4B,0x02,0x00,0x00,0x00,0x00,0x6B,0x02,0x2C,0x01, 0x04,0x3E,0x1E,0x00,0x01,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xD7,0xFA,0x00,0x1C,0x63,0xF2,0x1B,0xE0,0x04,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xDD,0x4B,0x9D,0x05,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x32,0x8C,0x00,0x30,0x06,0xE2,0xDA,0x89, 0x00,0x00,0x00,0x00,0x6B,0x22,0xD6,0x76,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x11,0xF3,0x00,0x2C,0xDD,0xC3,0x3E,0x91,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, 0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xB6,0xFE,0x00,0x01

Hope you can help out.

-e

g3gg0 commented:
a) as I didnt check that stuff yet, I cannot tell anything.
b) i can add those extended SLIX2 features (statically) and supply some test releases
c) send me a tag and I can check if it behaves as the original.

Please note that the signature seems to be a ECC signature and emulating it (i.e. generating matching it to the emulated UID) is nearly impossible.
Asymmetric ciphers are designed to not be cracked by hobbyists btw :)

NXP AN11600 describes how to verify the signature, but it is restricted.
**If anyone has it, feel free to mail  iso15693<at>g3gg0.de**

eychei commented:
Thanks.
It seems that if we just emulate the signature (memory) nothing needs to be computed.
The reader firmware will check if the signature is correct. 
Where can I send you a tag to? Will do asap.

-e

g3gg0 commented:
ah, found [this](https://github.com/free-dmo/free-dmo-stm32/blob/main/Src/main.c).

yes, that can be emulated easily i guess.

eychei commented:
I think one little bit is also missing.

The read password needs to be set to increment the counter. It is set for each tag differently. I can sniff the password for one of my tags and we could test on that data.

-e

g3gg0 commented:
if the code accepts any password, wouldn't that work as well?

g3gg0 commented:
the [code says](https://github.com/free-dmo/free-dmo-stm32/blob/main/Src/main.c#L492) so

eychei commented:
The firmware needs the correct password. It is generated for each UID. 
I do not have the generation scheme. 
The firmware is locked and I can not read it out. The STM32F072 is in RDP Level 2.
I do have a chipwhisperer but can not get it down to Level 1 or 0.

It may be that the counter will not work but the firmware may ignore that.
We could have the generate random number command send out the same number every time and the set password command always return a 1 (OK).

OH!
You are right. This is not needed. Just return the signal as true (success).

g3gg0 commented:
can you start a feature request / issue for this?

eychei commented:
Yes will do now. 

-e

eychei commented:
Added feature request here:

https://github.com/flipperdevices/flipperzero-firmware/issues/2781#issue-1761400779

semaf commented:
@g3gg0 

What is the point of "Can not save key file" after I try to save the read data (tonies)? Emulating is working. It still save it but get the error "Can not parse file".

Content of the .nfc file:
```
Filetype: Flipper NFC device
Version: 3
Nfc device type can be UID, Mifare Ultralight, Mifare Classic, FeliCa or ISO15693
Device type: ISO15693
```

I managed to read before with your fork that saved .nfc and .shd files but now the official firmware nor the rogue firmware is working. Not tested with your latest fork on github. Emulating on official firmware is not working but on Rogue firmware.

skotopes commented:
@semaf do you have proxmark ? can you send us card dump ?

g3gg0 commented:
Unfortunately I don't have my flipper at the moment.
Back then, the code did unlock and save the. nfc file immediately.
Seems something has changed and the saving code doesn't work anymore. Thus you get the error I guess?

semaf commented:
> @semaf do you have proxmark ? can you send us card dump ?

Yes I do but will have time the next days..

semaf commented:
> Unfortunately I don't have my flipper at the moment.
> 
> Back then, the code did unlock and save the. nfc file immediately.
> 
> Seems something has changed and the saving code doesn't work anymore. Thus you get the error I guess? 

The latest unleashed firmware version is working without any issue. Saving .nfc files.

bitboxer commented:
Is there a reason this is no longer on the dev branch of Flipper? This PR https://github.com/flipperdevices/flipperzero-firmware/pull/3050 removed it all?

team-orangeBlue commented:
Well, yes.
The NFC refactor blew up all of NFC. It now works for itself and nothing else.
That means, the app will work, but
* `nfc apdu` is gone
* Outside implementations, like nested/NFC magic are gone
* other firmwares (lol) just didn't merge this and when one did (MH-HM) the dev said his compiler exploded

So uhhh... yeah. But we still have ISO15693!

gornekich commented:
@bitboxer We are working on this functionality now. Will be ready soon

g3gg0 commented:
didn't follow the development since merge.
do you need support from my side?

gornekich commented:
@g3gg0 thanks, I think I will do it by my self and once it's done I will ask you to test it
